substrate/network-libp2p/src/custom_proto.rs

```diff
@@ -0,0 +1,270 @@
+	}
+
+	/// Returns the ID of the protocol.
+	pub fn id(&self) -> ProtocolId {
+		self.id
+	}
+
+	/// Returns the custom data that was passed to `new`.
+	pub fn custom_data(&self) -> &T {
+		&self.custom_data
+	}
+}
+
+impl<T, C, Maf> ConnectionUpgrade<C, Maf> for RegisteredProtocol<T>
+where C: AsyncRead + AsyncWrite + 'static,		// TODO: 'static :-/
+	  Maf: Future<Item = Multiaddr, Error = IoError> + 'static,		// TODO: 'static :(
+{
+	type NamesIter = VecIntoIter<(Bytes, Self::UpgradeIdentifier)>;
+	type UpgradeIdentifier = u8;		// Protocol version
+
+	#[inline]
+	fn protocol_names(&self) -> Self::NamesIter {
+		// Report each version as an individual protocol.
+		self.supported_versions.iter().map(|&(ver, _)| {
+			let num = ver.to_string();
+			let mut name = self.base_name.clone();
+			name.extend_from_slice(num.as_bytes());
+			(name, ver)
+		}).collect::<Vec<_>>().into_iter()
+	}
+
+	type Output = RegisteredProtocolOutput<T>;
+	type MultiaddrFuture = Maf;
+	type Future = future::FutureResult<(Self::Output, Self::MultiaddrFuture), IoError>;
+
+	fn upgrade(self, socket: C, protocol_version: Self::UpgradeIdentifier, _: Endpoint,
+		remote_addr: Maf) -> Self::Future
+	{
+		let packet_count = self.supported_versions
+			.iter()
+			.find(|&(v, _)| *v == protocol_version)
+			.expect("negotiated protocol version that wasn't advertised ; programmer error")
+			.1;
+
+		// This function is called whenever we successfully negotiated a protocol with a
+		// remote (both if initiated by us or by the remote)
+
+		// This channel is used to send outgoing packets to the custom_data for this open
+		// substream.
+		let (msg_tx, msg_rx) = mpsc::unbounded();
+
+		// Build the sink for outgoing network bytes, and the stream for incoming instructions.
+		// `stream` implements `Stream<Item = Message>`.
+		enum Message {
+			// Received data from the network.
+			RecvSocket(BytesMut),
+			// Data to send to the network.
+			// The packet_id must already be inside the `Bytes`.
+			SendReq(Bytes),
+			// The socket has been closed.
+			Finished,
+		}
+
+		let (sink, stream) = {
+			let framed = AsyncRead::framed(socket, VarintCodec::default());
+			let msg_rx = msg_rx.map(Message::SendReq)
+				.chain(stream::once(Ok(Message::Finished)))
+				.map_err(|()| unreachable!());
+			let (sink, stream) = framed.split();
+			let stream = stream.map(Message::RecvSocket)
+				.chain(stream::once(Ok(Message::Finished)));
+			(sink, msg_rx.select(stream))
+		};
+
+		let incoming = stream::unfold((sink, stream, false), move |(sink, stream, finished)| {
+			if finished {
+				return None;
+			}
+
+			Some(stream
+				.into_future()
+				.map_err(|(err, _)| err)
+				.and_then(move |(message, stream)| {
+					match message {
+						Some(Message::RecvSocket(mut data)) => {
+							// The `data` should be prefixed by the packet ID, therefore an
+							// empty packet is invalid.
+							if data.is_empty() {
+								debug!(target: "sub-libp2p", "ignoring incoming packet \
```

Comment:
still got some really heavily indented lines in this somewhat mammoth block. looks like it's a closure within a closure so may not be easy to refactor, but would consider it...